### ISSUE TITLE
Allow managing NVDA settings on secure screens

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#2226]
**Reported by KevanGC on 2012-04-06 08:19**
Currently, there's no very effective way of changing any settings for NVDA while in secure mode. The only option for most users is to use the "Use currently saved settings on the logon and other secure screens" button in the general settings dialog. However this button warns users about custom plugins being used in secure mode and people probably might feel discouraged from copying. And it's probably not safe using such plugins in secure mode, as they're security risks.

Also, it may take a long time copying settings and plugins over to secure mode, especially if one has lots of stuff in their %appdata%\NVDA folder. While copying, NVDA is completely unresponsive and freezes. It doesn't even give status indicators!

A better way of changing secure mode settings needs to be created. Perhaps you can only allow eSpeak to be used with no custom plugins. Many people including me don't want to hear slow eSpeak and have to use the desktop layout on a laptop in secure mode. I'm not griping, only trying to suggest something better.

Thanks for all your hard work put into NVDA!

**Comment by bhavyashah** [comment]
Since we have no means of contacting the original author of this ticket and no log or other diagnostic information to replicate the reported freeze, I suggest closing this ticket. @ehollig

**Comment by leonardder** [comment]
[**@bhavyashah**](https://github.com/bhavyashah) commented on [9 aug. 2017 17:06 CEST](https://github.com/nvaccess/nvda/issues/2226#issuecomment-321283959 "2017-08-09T15:06:31Z - Replied by Github Reply Comments"):
> I suggest closing this ticket. @ehollig

I tend to disagree with your reasoning here. While the freeze might not occur any more, I think the author states in general that there should be another, somewhat easier way to change settings for the logon screen, and I think he has a valid point.

**Comment by bhavyashah** [comment]
@leonardder According to @jcsteh's https://github.com/nvaccess/nvda/issues/2226#issuecomment-155296915, 'Allowing users to change settings while actually running on the secure desktop is a major security risk'. Therefore, I figured that Jamie was trying to clarify that even considering this suggestion was not possible due to security implications, and thus, only the reported freezes would be diagnosed further. As we both agree, the freezes may no longer be present.

**Comment by fernando-jose-silva** [comment]
I would also like to make a few suggestions:
The person who opened this call suggested that there might be an option to copy only the personal settings without the addons to the secure configuration of the system, this I believe could be interesting to avoid security risks. And if we want to make users even more impotent, it might be possible to choose which copy addons to copy when adding the addons to the secure configuration, since certain addons have a bad behavior on the logon screen, such as Addons that inform you of updates on the logon screen, and it is not possible to update the addon on this screen.
Another issue, if I choose to copy the settings along with the addons to the secure configuration, and if there is a large amount of data to be copied, a nvda screen is telling you to wait, and a continuous beep that does not indicate the percentage copied .
In this case it would be interesting that the nvda will inform you at least an estimate of the remaining time for the data to finish being copied.
Thank you.

**Comment by jcsteh** [comment]
#6305 covers selection of add-ons to copy. That would also allow choosing to copy no add-ons.

Beyond this, I don't know of a way we can achieve what's requested here without creating security concerns or potential problems in multi-user environments. @leonardder, unless you have other suggestions on how this could be done, I think this hsould be closed as won't fix (but noting that #6305 solves at least part of this).

**Comment by leonardder** [comment]
[**@jcsteh**](https://github.com/jcsteh) commented on [10 aug. 2017 01:57 CEST](https://github.com/nvaccess/nvda/issues/2226#issuecomment-321413074 "2017-08-09T23:57:43Z - Replied by Github Reply Comments"):
> Beyond this, I don't know of a way we can achieve what's requested here without creating security concerns or potential problems in multi-user environments. @leonardder, unless you have other suggestions on how this could be done...

When #577 is merged, it might be an option to allow the user to open the new NVDA settings dialog for the secure configuration, after gaining administrative privileges of course. In that case, we need to make sure only braille displays and speech synthesizers show up that are available in the secure screen config.

**Comment by ExplorerSunil** [comment]
Hi
   In some corporate settings, it makes sense to let user allow some configuration settings for secure screens or secure mode. E.g. my employer has enabled NVDA only in secured mode, and the startup command look like (so nvda is always on secure screens):

$ nvda.exe -r --secure --log-level=100 --disable-addons

   Therefore no configuration settings are saved. Each time NVDA crashes or needs to be restarted, I need to change the voice rate, pitch, synthesizer etc which is very cumbersome. 
So I think there should be a way to save these configurations separate from whether addons are allowed or not in these settings. Perhaps that would entail classifying the current settings into two parts one those should be saved for even secured mode and one shouldn't be.
   There isn't any option in secure mode for "use currently saved settings on logon and other secured screens" in general tab in NVDA settings. So, that trick doesn't work at all.

@Leonardder @nvdaes @josephsl

**Comment by josephsl** [comment]
Hi, it is intentional – users should not tamper with secure screens. One way to get around this is configure the setting you want while logged on, and then use copy settings button in general settings panel to copy the settings you want. Thanks.

**Comment by lukaszgo1** [comment]
It  is already possible to disable add-ons by supplying --disable-addons parameter. without providing --secure. What is not possible however, is to disable all riskier part of nNVDA (Python console, log viewer etc.) but still letting users save config. I believe this might solve your concerns.

**Comment by Adriani90** [comment]
It seems to be comfortable enough to copy the current settings from the NVDA user config to secure screen after the person is logged in. The part about add-ons is covered in #6305, so I am closing this in favor of #6305 as per https://github.com/nvaccess/nvda/issues/2226#issuecomment-321413074.

**Comment by CyrilleB79** [comment]
Sorry, I disagree.

Let's say I have in my user session speech rate set to 50 and screen curtain active. I have copied these parameters to logon screen settings.
Then I realize that for all users, it would be better to have speech rate set to 30 and screen curtain off.
Today, I need to configure these settings in my own session (and save them probably), copy it to logon screen, and then restore speech rate to 50 and screen curtain. That's not very straight forward.

As said before, we cannot allow users to modify settings directly from logon screen. Being able to open logon screen settings from a user session (with UAC elevation) as described in https://github.com/nvaccess/nvda/issues/2226#issuecomment-321450416 would be a nice solution.

It's worth noting that this pattern is already known by advanced Windows users, e.g. when you want to modify environment variables for all users or some other system-wide Windows settings.

Re-opening for discussion at least.

**Comment by ultrasound1372** [comment]
> As said before, we cannot allow users to modify settings directly from logon screen. Being able to open logon screen settings from a user session (with UAC elevation) as described in [#2226 (comment)](https://github.com/nvaccess/nvda/issues/2226#issuecomment-321450416) would be a nice solution.
> 
> It's worth noting that this pattern is already known by advanced Windows users, e.g. when you want to modify environment variables for all users or some other system-wide Windows settings.

Yes! I think this would be one of the best solutions to an issue we've been struggling with for years. The idea of a check list for which add-ons to copy has seemingly gone nowhere, and doing that with individual settings is finicky, plus add-ons that use extra configuration files would mess that up. I hadn't actually considered the case of wanting to have a different multi-user default, but that makes a lot of sense too. I think there should be some kind of mode we can launch NVDA in, with elevation, which we open on a user session and it lets us edit the secure config, sort of like a glorified profile. It would probably be simpler to actually restart NVDA into that mode rather than making it a profile, because it would be a special case all woven into profiles. Still means you have to configure things twice, but especially for the multi-user scenario that's what you'd want to do anyway in case they differ. So I'd launch this mode, probably from the general settings dialog, and I'd get an instance of NVDA using the secure config as its config, then I could install add-ons and change settings as I see fit. Then I'd just save configuration and exit, restarting back to my normal profile, and bam, new settings on secure screens, while actual secure mode is still immutable.
Should I open a separate feature request for this? I was planning to but had to look at prior issues to ensure I wasn't duplicating something, and found this last comment.

**Comment by SaschaCowley** [comment]
@ultrasound1372 feel free to create a new feature request for this

**Comment by ultrasound1372** [comment]
Alright, there's #18303. Hopefully it isn't merely closed as a duplicate.

**Comment by seanbudd** [comment]
Closing in favour of #18303 #18772 https://github.com/nvaccess/nvda/issues/6305
